### PR TITLE
Add AppConfig support for Lambda

### PR DESCRIPTION
**Pull request by mgorski-mg**
## Description
Add AWS AppConfig support for AWS Lambda. This functionality uses Lambda Extension to retrieve the config.

## Motivation and Context
It is extending the AWS AppConfig functionality to support AWS Lambdas properly.

## Testing
I created the AWS Lambda app and configure it with AppConfig using this new extension.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## License
- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by thebogusman** [comment]
I'm waiting impatiently for this to be merged!

**Comment by arturomonge** [comment]
Thanks for adding this, @mgorski-mg! Hopefully, this will get merged soon to start migrating our Lambda configs from SSM parameters to AppConfig, 👏👏👏.

**Comment by arturomonge** [comment]
@mgorski-mg I tested your `AddAppConfigForLambda` extension and is working perfectly, 👍. Thanks again!

I just have a comment about your recommendation to build `IConfiguration` in every invocation. I don't think this is desirable or practical or even necessary. I think a better recommendation is that if you wish for very fast reloading, you could use a very small `reloadAfter` value while using the `WaitForSystemsManagerReloadToComplete` method at the end of each request. This will achieve the same thing without having to constantly rebuild your configuration on each request and constantly calling the AWS AppConfig extension in each invocation.

In this function I just migrated to AppConfig, I configured the AWS AppConfig Lambda Extension to get a new configuration every 5 minutes. Then, I configured my Lambda function to reload the configuration after 1 minute, while adding the `WaitForSystemsManagerReloadToComplete` method at the end of each request. This way, I'm not either rebuilding `IConfiguration` or calling the AppConfig extension on every single invocation of the function (which could be a lot per minute). Instead, each lambda function reloads its configuration every 1 minute from the extension, and the extension checks for a new configuration every 5 minutes, which means that I'll get an updated configuration in my lambdas at most 6 minutes later (which in my use case, is good enough).

If you need very quick reloading, you could leave the AppConfig extension with the default poll interval (45 seconds) and set your lambda function to reload after 1 to 10 seconds (for example, depending on how fast you really need the reloading to happen). You will get a lot more calls to the AppConfig extension (which are super-quick) but without having to constantly rebuild `IConfiguration`.

In general, I don't think making a recommendation about the speed of reloading should be part of the readme, as this will vary from case to case. I think a single section for "Reloading in AWS Lambda" with just adding an example of using `AddAppConfigForLambda` with a `reloadAfter` parameter, and indicating that `WaitForSystemsManagerReloadToComplete` is needed, should be enough. Btw, the `AddSystemsManager` example doesn't show a `reloadAfter` parameter which I think it should.

**Comment by mgorski-mg** [comment]
I recommend it because reloading is done for us by Lambda Extension.
In all invocations after the cold start call to AppConfig last around 1ms because it's local communication.
But I think it's a good point, that maybe it shouldn't be recommended in the Readme, because it depends on what types of other configurations are you adding to the builder.
Let's wait for @normj and @philasmar and their opinion :)

Another topic is waiting for completion. 
`WaitForSystemsManagerReloadToComplete` is required only for Parameter Store I think because there you have internet communication and the whole config needs to be downloaded.
In AppConfig with Lambda Extension, you don't need to do any magic to wait for reload, because reloading is done on Lambda Extension level and in Lambda Handler is downloaded by local communication and takes around 1ms.

Btw. Thanks for testing my solution!!

**Comment by mgorski-mg** [comment]
@arturomonge I just realized that you are super right!
If we are building the `IConfiguration `in the AWS Lambda handler we are losing the caching at all because a new ClientId is created and a LastConfig version is also reset... Thank you so much! :)
Readme edited.

I also found one bug. When we are using Lambda Extension and the config is bigger than 2KB, the ContentLength property in the response from AppConfig(in this case from Lambda Extension) is 0, no idea why. 
Because of that, I change the logic of checking if the config needs to be parsed and cached in `IConfiguration` to verify the `ConfigurationVersion` property instead.

**Comment by mgorski-mg** [comment]
@arturomonge you are also right about waiting for reloading to be completed using `WaitForSystemsManagerReloadToComplete` method. Readme updated.
Thanks :)

**Comment by arturomonge** [comment]
@mgorski-mg Nice catch! I hadn't even thought about the ClientId being re-created every time `IConfiguration` was rebuilt, but glad you caught that.

About `WaitForSystemsManagerReloadToComplete`, yes, it's necessary if you are not rebuilding `IConfiguration` which now we know we shouldn't do for any reason, 😉. I see in your example you added it at the beginning of each invocation. In my case, I've added it at the end. This way, while the function is being executed, the reload is happening (which is an async process). If the reload finishes first, we don't have to wait for it as it will be completed by the time we call `WaitForSystemsManagerReloadToComplete`. But if the invocation is faster, then at the end of the invocation we will wait for the reload to complete. This way, we also give the OnChange token time to kick in while the invocation is occurring. I felt that if we call `WaitForSystemsManagerReloadToComplete` right at the beginning, the OnChange event might not have been triggered yet and we might miss it (sort of a race condition). It has been working great in a Lambda function already in production which I migrated to AWS AppConfig as a test.

Checking your commits, I had missed the use of `AWS_APPCONFIG_EXTENSION_PREFETCH_LIST`. Nice! I will add it right away to my functions to reduce cold start times, 👍.

File: src/Amazon.Extensions.Configuration.SystemsManager/Amazon.Extensions.Configuration.SystemsManager.csproj [hunk 1]
```diff
@@ -44,9 +44,9 @@
   </Choose>
 
   <ItemGroup>
-    <PackageReference Include="AWSSDK.AppConfig" Version="3.7.0.52" />
+    <PackageReference Include="AWSSDK.AppConfig" Version="3.7.0.61" />
     <PackageReference Include="AWSSDK.Extensions.NETCore.Setup" Version="3.7.1" />
-    <PackageReference Include="AWSSDK.SimpleSystemsManagement" Version="3.7.4.11" />
+    <PackageReference Include="AWSSDK.SimpleSystemsManagement" Version="3.7.5.7" />
     <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.0.*" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.*" />
```
Comment (damianh) [inline src/Amazon.Extensions.Configuration.SystemsManager/Amazon.Extensions.Configuration.SystemsManager.csproj:51]: Can we get rid of this reference?

Comment (mgorski-mg) [reply src/Amazon.Extensions.Configuration.SystemsManager/Amazon.Extensions.Configuration.SystemsManager.csproj:51]: @damianh, yes we can, but not in this PR. It is a different task 👍

Comment (damianh) [reply src/Amazon.Extensions.Configuration.SystemsManager/Amazon.Extensions.Configuration.SystemsManager.csproj:51]: Totally agree :)

**Comment by arturomonge** [comment]
@normj @philasmar There is an important bug in the support for AWS AppConfig that needs to be fixed, and it is fixed in this PR. It's when the AppConfig LambdaExtension returns 0 for content length if the config size is greater than 2Kb. I know you are full of things, but any possibility that you can give some attention to this PR? It would be extremely helpful. Thank you!

**Comment by normj** [comment]
I have created PR [#103 ](https://github.com/aws/aws-dotnet-extensions-configuration/pull/103) that consolidates this open PRs along with this change. Because of the breaking changes across the PRs including this one I am planning on a major version bump to 4.

Some significant rework was done on this PR in the consolidated PR to switch to the new `IAmazonAppConfigData` service client for how to get config data from AppConfig. I also switch to talking to the Lambda extension with a direct HttpClient calls.

I'm going to close this PR and track the work in #103. Please feel free to review that PR for the changes I made from this PR.

**Pull request closed without merging**